ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/CommandDeduplicationSpec.scala

```diff
@@ -44,165 +52,312 @@ class CommandDeduplicationSpec
+        commitContext.maximumRecordTime = Some(timestamp)
+        timestamp -> commitContext
+      }),
+      "normal-execution" -> ((dedupValueBuilder: Timestamp => Option[DamlStateValue]) => {
+        val dedupValue = dedupValueBuilder(timestamp)
+        val commitContext = createCommitContext(Some(timestamp), Map(aDedupKey -> dedupValue))
+        timestamp -> commitContext
+      }),
+    ).foreach { case (key, contextBuilder) =>
+      key should {
+        "continue if no deduplication entry could be found" in {
+          val (_, context) = contextBuilder(_ => None)
+
+          deduplicationStepContinues(context)
+        }
+
+        "continue if deduplication entry has no value set" in {
+          val (_, context) = contextBuilder(_ => Some(newDedupValue(identity)))
 
-      val actual =
-        deduplicateCommandStep(context, aTransactionEntrySummary)
+          deduplicationStepContinues(context)
+        }
+
+        "using deduplicate until" should {
 
-      actual match {
-        case StepContinue(_) => succeed
-        case StepStop(_) => fail()
+          "continue if record time is after deduplication time in case a deduplication entry is found" in {
+            val (_, context) = contextBuilder(timestamp =>
+              Some(
+                newDedupValue(
+                  _.setDeduplicatedUntil(buildTimestamp(timestamp.subtract(Duration.ofSeconds(1))))
+                )
+              )
+            )
+
+            deduplicationStepContinues(context)
+          }
+
+          "produce rejection log entry in case record time is on or before deduplication time" in {
+            for (
+              durationToAdd <- Iterable(
+                Duration.ZERO,
+                Duration.ofSeconds(1),
+              )
+            ) {
+              val (_, context) = contextBuilder(timestamp =>
+                Some(
+                  newDedupValue(
+                    _.setDeduplicatedUntil(buildTimestamp(timestamp.add(durationToAdd)))
+                  )
+                )
+              )
+              deduplicateStepHasTransactionRejectionEntry(context)
+            }
+          }
+        }
+
+        "using deduplication duration" should {
+          forAll(
+            Table[
+              String,
+              Timestamp => DamlCommandDedupValue.Builder => DamlCommandDedupValue.Builder,
+            ](
+              "identifier" -> "time setter",
+              "record time" -> ((timestamp: Timestamp) =>
+                (builder: DamlCommandDedupValue.Builder) =>
+                  builder.setRecordTime(buildTimestamp(timestamp))
+              ),
+              "max record time" -> ((timestamp: Timestamp) =>
+                (builder: DamlCommandDedupValue.Builder) =>
+                  builder.setRecordTimeBounds(
+                    buildPreExecutionDeduplicationBounds(timestamp, timestamp)
+                  )
+              ),
+            )
+          )(
```

Comment:
There are a lot of parentheses/brackets here and below. I think you should be able to improve readability with `{` instead of `(` here and using `case` for example. And I don't think you need to repeat the types.

Reply:
Did a try on this, let me know what you think. I do agree that it was a mess to read.

Reply:
LGTM, thanks!